src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+            #version 330 core
+            layout (location = 0) in vec3 aPos;
+
+            void main()
+            {
+                gl_Position = vec4(aPos.x, aPos.y, aPos.z, 1.0);
+            }
+        )";
+    static constexpr auto fallback_fragment_source = R"(
+            #version 330 core
+            out vec4 FragColor;
+            
+            void main()
+            {
+                FragColor = vec4(255.0f, 0.0f, 255.0f, 1.0f);
+            }    
+        )";
+
+    [[nodiscard]] constexpr std::string_view ShaderProgram::get_name_from_type(Type const type) {
+        switch (type) {
+            case Type::Vertex:
+                return "VERTEX";
+            case Type::Geometry:
+                return "GEOMETRY";
+            case Type::Fragment:
+                return "FRAGMENT";
+        }
+        return "INVALID";
+    }
+
+    [[nodiscard]] GLuint ShaderProgram::compile(Type const type, std::string_view const source) {
+        // compile
+
+        auto error_message = [&](GLuint id) -> std::string {
+            auto message = std::string{};
+            auto len = GLint{};
+            glGetShaderiv(id, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow_cast<usize>(len - 2));
+            glGetShaderInfoLog(id, len - 1, nullptr, message.data());
+            return message;
+        };
+        auto compile_single = [&](GLuint& id, char const* single_source) -> GLint {
+            id = glCreateShader(sly::to_underlying(type));
+            glShaderSource(id, 1, &single_source, nullptr);
+            glCompileShader(id);
+            auto success = GLint{};
+            glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+            return success;
+        };
+
+        auto id = GLuint{};
+        auto success = compile_single(id, source.data());
+
+        if (not success) {
+
+            auto message = error_message(id);
+            spdlog::critical("ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}", get_name_from_type(type), message);
+            glDeleteShader(id);
+
+            spdlog::info("SHADER::{}::TRY_FALLBACK_COMPILATION", get_name_from_type(type));
+            switch (type) {
+                case Type::Vertex:
+                    success = compile_single(id, fallback_vertex_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
```

Comment:
In my opinion, we should not handle the error if the fallback shader cannot be compiled. If this really happens, the whole program is unusable. So maybe just throw a `std::runtime_error` and let the whole machine ![](https://media1.giphy.com/media/oe33xf3B50fsc/giphy.gif?cid=ecf05e473031kgbf6kohcjukqstqmegsvixmg0t3zijrid17&ep=v1_gifs_search&rid=giphy.gif&ct=g)